File: R/pw_outlier.R
#' @title pw_outlier
#' @name pw_outlier
#' @description Pairwise outlier removal of replicate samples within analytical batches. \cr
#' This is useful for example to identify technical errors, particularly when there is not extensive replication among samples to conduct more conventional outlier detection.
#' @param df The dataframe containing peak data. At minimum it should contain columns labeled: name, sample, batch, compound, area, rep, rep_tech \cr
#' Additional columns are OK, but will not be used.
#' @param qc.label \code{character()} Label designating the QC sample in the sample column of df. These will not be subjected to pairwise outlier detection since they can be useful downstream, for example with signal drift correction.
#' @param n.cores \code{numeric()} The number of cores to be used for processing if being run on a multi-core machine
#' @param mad_threshold \code{numeric()} The median absolute deviation (MAD) threshold to be used when identifying potential outliers
#' @param pw_threshold \code{numeric()} The pairwise outlier difference threshold. Should be between 0 and 1. Default is set to 0.95, meaning 5% of the data will be identified as a potential outlier.
#' @param peak_shrinkage \code{logical()} For samples surpassing the MAD threshold, should they be shrunk towards the distribution (TRUE) or completely removed from the analysis (FALSE). \cr
#' If set to TRUE (default), values are shrunk towards the distribution of samples and will maintain the same rankings of samples while reducing the overall distribution skew during the pairwise elimination calculation.
#' @param grouping_factor The column label containing the grouping factor from which pairwise differences will be calculated. Default is batch column.
#' @param return_plot \code{logical()} Should a density plot be returned showing the pairwise difference threshold distributions per compound. Default is FALSE.
#' @param plot_name \code{character()} If return_plot is TRUE, what is the name of the output .pdf file. This will be saved to the current working directory.
#' @return \code{list()} containing:
#' \itemize{
#' \item df (original input data)
#' \item df_cleaned (pairwise outlier cleaned data).
#' \item df_rm (samples removed by the pairwise outlier elimination).
#' }
#' @import dplyr tidyr
#' @importFrom data.table fread
#' @importFrom parallel mclapply
#' @export
#' @examples
#' pw_out = pw_outlier(
#' df = dat,
#' qc.label = "Pool",
#' n.cores = 2,
#' mad_threshold = 3,
#' pw_threshold = 0.95,
#' peak_shrinkage = TRUE,
#' grouping_factor = "batch",
#' return_plot = FALSE,
#' plot_name = "pw_outlier_plot")
#'
#' list2env(pw_out ,.GlobalEnv)
pw_outlier <- function(
  df = NULL,
  n.cores = 1,
  mad_threshold = 3,
  pw_threshold = 0.95,
  peak_shrinkage = TRUE,
  grouping_factor = "batch",
  return_plot = FALSE,
  plot_name = "pw_outlier_plot",
  qc.label = "Pool"){
  # check input df colnames -------------------------------------------------
  c_names = colnames(df)
  c_names_need = c("name","sample","batch","compound","area","rep","rep_tech")
  c_names_check = c_names_need[!c_names_need%in%c_names]
  if (length(c_names_check)>=1) {
    print(paste0("Please check your input df to make sure it contains all necessary columns... ",c_names_check, " missing"))
    stop()
  }
  m = df
  gp = grouping_factor
  gplab = m[,gp] %>%
    pull(all_of(gp))
  gplab = factor(gplab, levels = unique(gplab))
  m = m %>%
    mutate(g = gplab)
  # make lists to hold output -----------------------------------------------
  samps_rm = list()
  samps_plot = list()
  # additional df calculations by compound ----------------------------------
  m = m %>%
    group_by(sample, compound, rep_tech, g) %>%
    mutate(geno_tmp = paste0("TR",rep_tech,"_", sample), .before = 1) %>%
    ungroup() %>%
    group_by(geno_tmp, compound, rep_tech, g) %>%
    mutate(tmp_rep = 1:n(),
           uid = paste0(geno_tmp, "_", tmp_rep, "_", g, "_",compound)) %>%
    ungroup() %>%
    group_by(g, compound) %>%
    mutate(my_rank = rank(area, na.last = "keep", ties.method = "average"),
           my_quant = qnorm(my_rank/(n()+1)),
           my_mad = mad(area, na.rm = TRUE),
           my_thresh = all_of(mad_threshold)*my_mad + median(area, na.rm = TRUE),
           area_tmp = ifelse(area>=my_thresh, NA,area),
           area_shrink = ifelse(area>=my_thresh, max(area_tmp, na.rm = TRUE)+abs(my_quant),area))
  if (peak_shrinkage) {
    m = m %>%
      mutate(area_og = area,
             area = area_shrink)
  }else{
    m = m %>%
      mutate(area_og = area,
             area = area_tmp)
  }
  mog = m
  m = m %>%
    filter(!sample==all_of(qc.label))
  m_na = m %>%
    filter(is.na(area))
  m = m %>%
    drop_na(area)
  ## Change option to supress warning message when summarizing
  options(dplyr.summarise.inform = FALSE)
  ## Compounds in data
  integrated_compounds = sort(unique(m$compound))
  ## Calculate pairwise difference between genotypes separated by batch or other grouping factor.
  pw_fn = function(ii){
    all_compounds = integrated_compounds[ii]
    for (i in seq_along(all_compounds)) {
      ## Make an outlier shrikage while maintaining rankings
      my_df = m %>%
        filter(compound==all_of(all_compounds[i]))
      my_batches = unique(my_df$g)
      diff_vect = list()
      most_diff = list()
      plot_dat = list()
      for (j in seq_along(my_batches)) {
        current_batch = my_df %>%
          filter(g == all_of(my_batches[j]))
        my_genos = unique(current_batch$geno_tmp)
        diff_vect1 = list()
        most_diff1 = list()
        for (k in seq_along(my_genos)) {
          current_geno = current_batch %>%
            filter(geno_tmp==all_of(my_genos[k]))
          current_geno_name = unique(current_geno$geno_tmp)
          n_obs = nrow(current_geno)
          if (n_obs<3) {
            most_diff1[[k]] = NULL
            diff_vect1[[k]] = NULL
            # print(paste0("Genotype: ",my_genos[k], " in ", my_batches[j], " has less than three obs for compound: ", all_compounds[i]))
          }else{
            geno_diffs = as_tibble(as.matrix(dist(current_geno$area, method = "manhattan",upper = TRUE))) %>%
              rowwise() %>%
              mutate(sum_diffs = sum(c_across(cols = everything())))
            most_diff1[[k]] = as_tibble(geno_diffs) %>%
              mutate(tmp_rep = 1:n()) %>%
              mutate(geno_tmp = all_of(current_geno_name)) %>%
              slice_max(., order_by = sum_diffs, prop = 0.34) %>%
              mutate(compound = all_of(all_compounds[i])) %>%
              mutate(g = all_of(my_batches[j])) %>%
              mutate(uid = paste0(geno_tmp, "_", tmp_rep, "_", g))
            diff_vect1[[k]] = as_tibble(as.numeric(dist(current_geno$area, method = "manhattan"))) %>%
              mutate(compound = all_of(all_compounds[i])) %>%
              mutate(g = all_of(my_batches[j]))
          }
        }
        diff_vect[[j]] = bind_rows(diff_vect1) %>%
          group_by(compound, g) %>%
          summarise(thresh = quantile(value, pw_threshold, na.rm = TRUE))
        plot_dat[[j]] = bind_rows(diff_vect1) %>%
          left_join(., diff_vect[[j]], by = c("compound", "g"))
        most_diff[[j]] = bind_rows(most_diff1) %>%
          select(-c(geno_tmp, sum_diffs, tmp_rep)) %>%
          pivot_longer(!c("uid", "compound", "g"),names_to = "variable") %>%
          left_join(., diff_vect[[j]], by = c("compound", "g")) %>%
          filter(value>=thresh)
      }
      samps_rm[[i]] = bind_rows(most_diff)
      samps_plot[[i]] = bind_rows(plot_dat)
    }
    return(list(samps_rm = samps_rm,samps_plot = samps_plot))
  }
  pw_results = parallel::mclapply(seq_along(integrated_compounds), pw_fn, mc.cores = n.cores)
  samps_rm1 = lapply(pw_results,"[[", 1)
  samps_rm1 = bind_rows(samps_rm1) %>%
    select(-c(variable, value)) %>%
    distinct() %>%
    mutate(uid = paste0(uid, "_", compound))
  samps_plot1 = lapply(pw_results,"[[", 2)
  samps_plot1 = bind_rows(samps_plot1)
  if(return_plot){
    pdf(paste0(plot_name,".pdf"),width = 10,height = 10)
    for (i in seq_along(integrated_compounds)) {
      comp_thresh = samps_plot1 %>%
        filter(compound==all_of(integrated_compounds[i])) %>%
        select(-c(value)) %>%
        distinct()
      p = samps_plot1 %>%
        filter(compound==all_of(integrated_compounds[i])) %>%
        ggplot(., aes(x=value)) +
        geom_density(alpha=0.2) +
        facet_wrap(~ g) +
        geom_vline(aes(xintercept=thresh), colour = "#2c7fb8")+
        geom_label(data = comp_thresh, mapping = aes(x = -Inf, y = -Inf, label = paste0(pw_threshold, " threshold: ", round(thresh, 1) )), hjust = -0.79, vjust = -1.8, colour = "black")+
        theme_classic()+
        theme(
          axis.text.y = element_text(size = 12,face = "bold", color = "black"),
          axis.title.y = element_text(size = 12, color = "black", face = "bold"),
          axis.title.x = element_text(size = 12, color = "black", face = "bold"),
          strip.text.x = element_text(size = 9, color = "black", face = "bold"),
          plot.title = element_text(hjust = 0.5,color = "black", face = "bold"),
          legend.key.width = unit(2,"cm"),
          legend.position="top",
          legend.box="vertical",
          legend.margin=margin(),
          legend.justification='center',
          # legend.title = element_blank(),
          legend.direction='horizontal'
        )+
        labs(x = "Replication pairwise differences", y = "Density",title = integrated_compounds[i])
      print(p)
    }
    dev.off()
  }
  # outputs to return -------------------------------------------------------
  dat_removed = m %>%
    ungroup() %>%
    filter(uid%in%samps_rm1$uid) %>%
    bind_rows(m_na) %>%
    mutate(area = area_og)
  m1 = mog %>%
    ungroup() %>%
    filter(!uid%in%dat_removed$uid) %>%
    mutate(area = area_og) %>%
    select(all_of(c_names))
  dat_removed = dat_removed %>%
    select(all_of(c_names))
  dd <- c(" ", " -------------- ", " Done! ", " --------------")
  cat(dd, sep = "\n")
  return(list(df = df, df_cleaned = m1, df_rm = dat_removed))
}

File: pseudoDrift/pw_outlier.py
import pandas as pd
import numpy as np
from scipy.spatial import distance
from scipy.stats import rankdata, norm
from scipy.stats import median_abs_deviation as mad
import seaborn as sns
import matplotlib.pyplot as plt

# Function to compute median absolute deviation
def compute_mad(arr):
    med = np.median(arr)
    return np.median(np.abs(arr - med))

## Main function
def pw_outlier(df=None, n_cores=1, mad_threshold=3, pw_threshold=0.95, 
               peak_shrinkage=True, grouping_factor="batch", samps_exclude="QC"):
    
    # Helper function to calculate most differing observations
    def get_most_diff(pairwise_diffs, current_geno, threshold_prop=0.05):
        num_samples = current_geno.shape[0]
        geno_diffs = pd.DataFrame(pairwise_diffs.reshape((num_samples, num_samples)), columns=current_geno['uid'].values, index=current_geno['uid'].values)
        geno_diffs['sum_diffs'] = geno_diffs.sum(axis=1)
        top_diff_indices = geno_diffs['sum_diffs'].nlargest(int(len(geno_diffs) * threshold_prop)).index
        most_diff = current_geno[current_geno['uid'].isin(top_diff_indices)].copy()
        most_diff['sum_diffs'] = geno_diffs['sum_diffs'].loc[top_diff_indices].values
        return most_diff
    
    # Check input df colnames
    c_names = df.columns
    c_names_need = ["name", "sample", "batch", "compound", "area", "rep", "rep_tech"]
    c_names_check = [col for col in c_names_need if col not in c_names]
    
    if len(c_names_check) >= 1:
        raise ValueError(f"Please check your input df to make sure it contains all necessary columns... {c_names_check} missing")
    
    m = df.copy()
    m[grouping_factor] = m[grouping_factor].astype('category')
    m['g'] = m[grouping_factor].cat.codes
    
    # Additional df calculations by compound
    m['geno_tmp'] = "TR" + m['rep_tech'].astype(str) + "_" + m['sample'].astype(str)
    m['tmp_rep'] = m.groupby(['geno_tmp', 'compound', 'rep_tech', 'g']).cumcount() + 1
    m['uid'] = m['geno_tmp'] + "_" + m['tmp_rep'].astype(str) + "_" + m['g'].astype(str) + "_" + m['compound']
    
    grouped = m.groupby(['g', 'compound'])
    m['my_rank'] = grouped['area'].transform(lambda x: rankdata(x))
    m['my_quant'] = norm.ppf(m['my_rank'] / (grouped['area'].transform('size') + 1))
    m['my_mad'] = grouped['area'].transform(lambda x: compute_mad(x))
    m['my_thresh'] = mad_threshold * m['my_mad'] + grouped['area'].transform('median')
    
    m['area_tmp'] = np.where(m['area'] > m['my_thresh'], np.nan, m['area'])
    m['area_shrink'] = np.where(m['area'] > m['my_thresh'], grouped['area_tmp'].transform('max') + m['my_quant'], m['area'])
    
    if peak_shrinkage:
        m['area_og'] = m['area']
        m['area'] = m['area_shrink']
    else:
        m['area_og'] = m['area']
        m['area'] = m['area_tmp']
    
    m['uid_compound'] = m['uid'] + "_" + m['compound']
    mog = m.copy()
    m = m[~m['sample'].isin([samps_exclude])]
    
    # Compounds in data
    integrated_compounds = m['compound'].unique()
    samps_rm = []
    samps_plot = []
    
    # Loop through compounds
    for compound in integrated_compounds:
        my_df = m[m['compound'] == compound]
        my_batches = my_df['g'].unique()
        diff_vect = []
        most_diff = []
        plot_dat = []
        
        for batch in my_batches:
            current_batch = my_df[my_df['g'] == batch]
            my_genos = current_batch['geno_tmp'].unique()
            diff_vect1 = []
            most_diff1 = []
            
            for geno in my_genos:
                current_geno = current_batch[current_batch['geno_tmp'] == geno]
                current_geno_name = geno
                
                if len(current_geno) < 3:
                    most_diff1.append(None)
                    diff_vect1.append(None)
                else:
                    pairwise_diffs = distance.squareform(distance.pdist(current_geno['area'].values.reshape(-1, 1), 'cityblock'))
                    most_diff1.append(get_most_diff(pairwise_diffs, current_geno))
                    diff_vect1.append(pd.DataFrame({
                        'value': pairwise_diffs.flatten(),
                        'compound': compound,
                        'g': batch
                    }))
            
            if len(diff_vect1) == 0:
                diff_vect.append(None)
                plot_dat.append(None)
                most_diff.append(None)
            else:
                diff_vect.append(pd.concat(diff_vect1))
                most_diff.append(pd.concat(most_diff1))

        samps_rm.append(pd.concat(most_diff))
        samps_plot.append(pd.concat(diff_vect))
    
    # Calculate the threshold for each batch and compound in each batch
    thresholds = []
    for df in samps_plot:
        if df is not None:
            thresh = df.groupby(['compound', 'g'])['value'].quantile(pw_threshold).reset_index()
            thresholds.extend(thresh.to_dict(orient='records'))
    
    # Filter out outliers
    outliers = pd.concat(samps_rm)
    outliers = outliers.drop_duplicates(subset=['uid', 'compound'])
    outliers['uid_compound'] = outliers['uid'] + "_" + outliers['compound']
    m1 = mog[~mog['uid_compound'].isin(outliers['uid_compound'])]
    m1['area'] = m1['area_og']
    m1 = m1[c_names]
    
    return {
        'df_cleaned': m1,
        'df_rm': outliers[c_names],
        'thresholds': thresholds
    }


# Generate larger sample dataset to test function
n_samples = 1000
batches = ['batch_' + str(i) for i in range(1, 6)]
compounds = ['compound_' + str(i) for i in range(1, 4)]
reps = [1, 2, 3]
rep_techs = ['tech_' + str(i) for i in range(1, 4)]

df_sample = pd.DataFrame({
    'name': ['name_' + str(i) for i in range(n_samples) for _ in compounds for _ in reps for _ in rep_techs],
    'sample': ['sample_' + str(i) for i in range(n_samples) for _ in compounds for _ in reps for _ in rep_techs],
    'batch': np.random.choice(batches, n_samples*len(compounds)*len(reps)*len(rep_techs)),
    'compound': np.tile(np.repeat(compounds, len(reps)*len(rep_techs)), n_samples),
    'rep': np.tile(np.repeat(reps, len(rep_techs)), n_samples*len(compounds)),
    'rep_tech': np.tile(rep_techs, n_samples*len(compounds)*len(reps)),
})

# Simulate area values for each compound with some outliers
compound_means = {'compound_1': 100, 'compound_2': 200, 'compound_3': 50}
std_dev = 10

for compound, mean_value in compound_means.items():
    mask = df_sample['compound'] == compound
    df_sample.loc[mask, 'area'] = np.random.normal(mean_value, std_dev, mask.sum())

    # Introducing some outliers
    outlier_indices = np.random.choice(df_sample[mask].index, size=int(0.05 * mask.sum()), replace=False)
    df_sample.loc[outlier_indices, 'area'] = mean_value + np.random.choice([50, -50], len(outlier_indices))

df_sample.head()

# Test the function
my_results = pw_outlier(df_sample)

## Notes: There is a slight bug that needs to be fixed in plotting function below

""" # Function for plotting
def generate_and_display_plots(df, thresholds):
    integrated_compounds = df['compound'].unique().tolist()

    for compound in integrated_compounds:
        compound_diffs = df[df['compound'] == compound]['area'].tolist()
        
        # Plot pairwise differences for the compound
        plt.figure(figsize=(10, 5))
        sns.kdeplot(compound_diffs, shade=True, color='lightblue')
        for batch_info in thresholds:
            if batch_info['compound'] == compound:
                plt.axvline(x=batch_info['threshold'], color='red', linestyle='--')
        plt.title(f"Compound: {compound}")
        plt.xlabel("Replication pairwise differences")
        plt.ylabel("Density")
        plt.grid(True, which='both', linestyle='--', linewidth=0.5)
        plt.tight_layout()
        plt.show()

## Test the plotting function
generate_and_display_plots(my_results['df_cleaned'], my_results['thresholds']) """
